File: demo-app/app/src/javaApp/java/com/launchkey/android/authenticator/demo/ui/fragment/AppConfigsViewModel.java
package com.launchkey.android.authenticator.demo.ui.fragment;

import android.content.Context;

import androidx.annotation.NonNull;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.SavedStateHandle;
import androidx.lifecycle.ViewModel;

import com.launchkey.android.authenticator.demo.R;
import com.launchkey.android.authenticator.sdk.core.auth_method_management.AuthMethod;
import com.launchkey.android.authenticator.sdk.core.authentication_management.AuthenticatorConfig;
import com.launchkey.android.authenticator.sdk.core.authentication_management.AuthenticatorManager;
import com.launchkey.android.authenticator.sdk.core.authentication_management.Device;
import com.launchkey.android.authenticator.sdk.core.authentication_management.event_callback.UnlinkDeviceEventCallback;
import com.launchkey.android.authenticator.sdk.ui.AuthenticatorUIConfig;
import com.launchkey.android.authenticator.sdk.ui.AuthenticatorUIManager;
import com.launchkey.android.authenticator.sdk.ui.internal.viewmodel.SingleLiveEvent;
import com.launchkey.android.authenticator.sdk.ui.theme.AuthenticatorTheme;

public class AppConfigsViewModel extends ViewModel {
    @NonNull
    private final AuthenticatorManager authenticatorManager;

    @NonNull
    private final AuthenticatorUIManager authenticatorUIManager;

    @NonNull
    private final SavedStateHandle savedStateHandle;

    @NonNull
    private String configsSdkKey;
    @NonNull
    public String getConfigsSdkKey() { return configsSdkKey; }
    public void setConfigsSdkKey(@NonNull final String configsSdkKey) {
        this.savedStateHandle.set("configsSdkKey", configsSdkKey);
        this.configsSdkKey = configsSdkKey;
    }
    private boolean configsSslPinning;
    public boolean getConfigsSslPinning() { return configsSslPinning; }
    public void setConfigsSslPinning(final boolean configsSslPinning) {
        this.savedStateHandle.set("configsSslPinning", configsSslPinning);
        this.configsSslPinning = configsSslPinning;
    }
    private boolean configsPc;
    public boolean getConfigsPc() { return configsPc; }
    public void setConfigsPc(final boolean configsPc) {
        this.savedStateHandle.set("configsPc", configsPc);
        this.configsPc = configsPc;
    }
    private boolean configsCc;
    public boolean getConfigsCc() { return configsCc; }
    public void setConfigsCc(final boolean configsCc) {
        this.savedStateHandle.set("configsCc", configsCc);
        this.configsCc = configsCc;
    }
    private boolean configsW;
    public boolean getConfigsW() { return configsW; }
    public void setConfigsW(final boolean configsW) {
        this.savedStateHandle.set("configsW", configsW);
        this.configsW = configsW;
    }
    private boolean configsL;
    public boolean getConfigsL() { return configsL; }
    public void setConfigsL(final boolean configsL) {
        this.savedStateHandle.set("configsL", configsL);
        this.configsL = configsL;
    }
    private boolean configsFs;
    public boolean getConfigsFs() { return configsFs; }
    public void setConfigsFs(final boolean configsFs) {
        this.savedStateHandle.set("configsFs", configsFs);
        this.configsFs = configsFs;
    }
    private int configsDelayWearables;
    public int getConfigsDelayWearables() { return configsDelayWearables; }
    public void setConfigsDelayWearables(final int configsDelayWearables) {
        this.savedStateHandle.set("configsDelayWearables", configsDelayWearables);
        this.configsDelayWearables = configsDelayWearables;
    }
    private int configsDelayLocations;
    public int getConfigsDelayLocations() { return configsDelayLocations; }
    public void setConfigsDelayLocations(final int configsDelayLocations) {
        this.savedStateHandle.set("configsDelayLocations", configsDelayLocations);
        this.configsDelayLocations = configsDelayLocations;
    }
    private int configsAuthfailure;
    public int getConfigsAuthfailure() { return configsAuthfailure; }
    public void setConfigsAuthfailure(final int configsAuthfailure) {
        this.savedStateHandle.set("configsAuthfailure", configsAuthfailure);
        this.configsAuthfailure = configsAuthfailure;
    }
    private int configsAutounlink;
    public int getConfigsAutounlink() { return configsAutounlink; }
    public void setConfigsAutounlink(final int configsAutounlink) {
        this.savedStateHandle.set("configsAutounlink", configsAutounlink);
        this.configsAutounlink = configsAutounlink;
    }
    private int configsAutounlinkwarning;
    public int getConfigsAutounlinkwarning() { return configsAutounlinkwarning; }
    public void setConfigsAutounlinkwarning(final int configsAutounlinkwarning) {
        this.savedStateHandle.set("configsAutounlinkwarning", configsAutounlinkwarning);
        this.configsAutounlinkwarning = configsAutounlinkwarning;
    }
    private boolean configsAllowchangesunlinked;
    public boolean getConfigsAllowchangesunlinked() { return configsAllowchangesunlinked; }
    public void setConfigsAllowchangesunlinked(final boolean configsAllowchangesunlinked) {
        this.savedStateHandle.set("configsAllowchangesunlinked", configsAllowchangesunlinked);
        this.configsAllowchangesunlinked = configsAllowchangesunlinked;
    }
    private AuthenticatorTheme authenticatorTheme;
    public void setAuthenticatorTheme(final AuthenticatorTheme authenticatorTheme) {
        this.authenticatorTheme = authenticatorTheme;
    }
    
    public AppConfigsViewModel(@NonNull final AuthenticatorManager authenticatorManager,
                               @NonNull final AuthenticatorUIManager authenticatorUIManager,
                               @NonNull final String sdkKey,
                               @NonNull final SavedStateHandle savedStateHandle) {
        this.authenticatorManager = authenticatorManager;
        this.authenticatorUIManager = authenticatorUIManager;
        this.savedStateHandle = savedStateHandle;

        final AuthenticatorConfig currentConfig = authenticatorManager.getConfig();
        if (this.savedStateHandle.get("configsSdkKey") != null) configsSdkKey = this.savedStateHandle.get("configsSdkKey");
        else configsSdkKey = sdkKey;
        if (this.savedStateHandle.get("configsSslPinning") != null) configsSslPinning = this.savedStateHandle.get("configsSslPinning");
        else configsSslPinning = currentConfig.isSslPinningRequired();
        if (this.savedStateHandle.get("configsPc") != null) configsPc = this.savedStateHandle.get("configsPc");
        else configsPc = currentConfig.isMethodAllowed(AuthMethod.PIN_CODE);
        if (this.savedStateHandle.get("configsCc") != null) configsCc = this.savedStateHandle.get("configsCc");
        else configsCc = currentConfig.isMethodAllowed(AuthMethod.CIRCLE_CODE);;
        if (this.savedStateHandle.get("configsW") != null) configsW = this.savedStateHandle.get("configsW");
        else configsW = currentConfig.isMethodAllowed(AuthMethod.WEARABLES);;
        if (this.savedStateHandle.get("configsL") != null) configsL = this.savedStateHandle.get("configsL");
        else configsL = currentConfig.isMethodAllowed(AuthMethod.LOCATIONS);;
        if (this.savedStateHandle.get("configsFs") != null) configsFs = this.savedStateHandle.get("configsFs");
        else configsFs = currentConfig.isMethodAllowed(AuthMethod.BIOMETRIC);;
        if (this.savedStateHandle.get("configsDelayWearables") != null) configsDelayWearables = this.savedStateHandle.get("configsDelayWearables");
        else configsDelayWearables = currentConfig.activationDelayWearablesSeconds();
        if (this.savedStateHandle.get("configsDelayLocations") != null) configsDelayLocations = this.savedStateHandle.get("configsDelayLocations");
        else configsDelayLocations = currentConfig.activationDelayLocationsSeconds();
        if (this.savedStateHandle.get("configsAuthfailure") != null) configsAuthfailure = this.savedStateHandle.get("configsAuthfailure");
        else configsAuthfailure = currentConfig.thresholdAuthFailure();
        if (this.savedStateHandle.get("configsAutounlink") != null) configsAutounlink = this.savedStateHandle.get("configsAutounlink");
        else configsAutounlink = currentConfig.thresholdAutoUnlink();
        if (this.savedStateHandle.get("configsAutounlinkwarning") != null) configsAutounlinkwarning = this.savedStateHandle.get("configsAutounlinkwarning");
        else configsAutounlinkwarning = currentConfig.thresholdAutoUnlinkWarning();

        final AuthenticatorUIConfig currentUIConfig = authenticatorUIManager.getConfig();
        if (this.savedStateHandle.get("configsAllowchangesunlinked") != null) configsAllowchangesunlinked = this.savedStateHandle.get("configsAllowchangesunlinked");
        else configsAllowchangesunlinked = currentUIConfig.areSecurityChangesAllowedWhenUnlinked();
    }

    public void finalizeChangesToConfigAndRestart(@NonNull final Context context) {
        _state.setValue(new State.ChangesLoading());
        if (configsSdkKey.trim().isEmpty()) {
            _state.setValue(new State.ChangesFailed("Key cannot be null or empty."));
            return;
        }

        // Let Authenticator SDK validate threshold arguments
        final AuthenticatorConfig config;
        final AuthenticatorUIConfig uiConfig;
        try {
            AuthenticatorConfig.Builder builder = new AuthenticatorConfig.Builder();
            disallowAuthMethod(builder, AuthMethod.PIN_CODE, configsPc);
            disallowAuthMethod(builder, AuthMethod.CIRCLE_CODE, configsCc);
            disallowAuthMethod(builder, AuthMethod.WEARABLES, configsW);
            disallowAuthMethod(builder, AuthMethod.LOCATIONS, configsL);
            disallowAuthMethod(builder, AuthMethod.BIOMETRIC, configsFs);
            config = builder
                    .activationDelayWearable(configsDelayWearables)
                    .activationDelayLocation(configsDelayLocations)
                    .thresholdAuthFailure(configsAuthfailure)
                    .thresholdAutoUnlink(configsAutounlink)
                    .thresholdAutoUnlinkWarning(configsAutounlinkwarning)
                    .keyPairSize(AuthenticatorConfig.Builder.KEYSIZE_MINIMUM)
                    .sslPinning(configsSslPinning)
                    .build();

            uiConfig = new AuthenticatorUIConfig.Builder()
                    .allowSecurityChangesWhenUnlinked(configsAllowchangesunlinked)
                    .theme(authenticatorTheme)
                    .build(context.getApplicationContext());
        } catch (IllegalArgumentException e) {
            _state.setValue(new State.ChangesFailed(e.getMessage()));
            return;
        }

        if (authenticatorManager.isDeviceLinked()) {
            // Force-unlink to clear any data before re-initializing
            authenticatorManager.unlinkDevice(null, new UnlinkDeviceEventCallback() {
                @Override
                public void onSuccess(final @NonNull Device device) {
                    authenticatorManager.initialize(config);
                    authenticatorUIManager.initialize(uiConfig);
                    _state.setValue(new State.ChangesDone(configsSdkKey));
                }

                @Override
                public void onFailure(final @NonNull Exception e) {
                    authenticatorManager.initialize(config);
                    authenticatorUIManager.initialize(uiConfig);
                    _state.setValue(new State.ChangesDone(configsSdkKey));
                }
            });
        } else {
            authenticatorManager.initialize(config);
            authenticatorUIManager.initialize(uiConfig);
            _state.setValue(new State.ChangesDone(configsSdkKey));
        }
    }

    private static void disallowAuthMethod(final @NonNull AuthenticatorConfig.Builder configBuilder, final @NonNull AuthMethod authMethod, final boolean allow) {
        if (!allow) {
            configBuilder.disallowAuthMethod(authMethod);
        }
    }

    @NonNull
    private final MutableLiveData<State> _state = new SingleLiveEvent<>();
    public final LiveData<State> state = _state;

    public abstract static class State {
        public static final class ChangesFailed extends State {
            @NonNull
            public final String failure;
            ChangesFailed(@NonNull final String failure) { this.failure = failure; }
        }
        public static final class ChangesLoading extends State {}
        public static final class ChangesDone extends State {
            @NonNull
            public final String sdkKey;
            ChangesDone(@NonNull final String sdkKey) {
                this.sdkKey = sdkKey;
            }
        }
    }
}

File: demo-app/app/src/kotlinApp/java/com/launchkey/android/authenticator/demo/ui/fragment/AppConfigsViewModel.kt
package com.launchkey.android.authenticator.demo.ui.fragment

import android.content.Context
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import com.launchkey.android.authenticator.sdk.core.auth_method_management.AuthMethod
import com.launchkey.android.authenticator.sdk.core.authentication_management.AuthenticatorConfig
import com.launchkey.android.authenticator.sdk.core.authentication_management.AuthenticatorManager
import com.launchkey.android.authenticator.sdk.core.authentication_management.Device
import com.launchkey.android.authenticator.sdk.core.authentication_management.event_callback.UnlinkDeviceEventCallback
import com.launchkey.android.authenticator.sdk.ui.AuthenticatorUIConfig
import com.launchkey.android.authenticator.sdk.ui.AuthenticatorUIManager
import com.launchkey.android.authenticator.sdk.ui.internal.viewmodel.SingleLiveEvent
import com.launchkey.android.authenticator.sdk.ui.theme.AuthenticatorTheme

class AppConfigsViewModel(private val authenticatorManager: AuthenticatorManager,
                          private val authenticatorUIManager: AuthenticatorUIManager,
                          sdkKey: String,
                          private val savedStateHandle: SavedStateHandle) : ViewModel() {
    private var configsSdkKey: String
    fun getConfigsSdkKey(): String {
        return configsSdkKey
    }

    fun setConfigsSdkKey(configsSdkKey: String) {
        savedStateHandle.set("configsSdkKey", configsSdkKey)
        this.configsSdkKey = configsSdkKey
    }

    private var configsSslPinning = false
    fun getConfigsSslPinning(): Boolean {
        return configsSslPinning
    }

    fun setConfigsSslPinning(configsSslPinning: Boolean) {
        savedStateHandle.set("configsSslPinning", configsSslPinning)
        this.configsSslPinning = configsSslPinning
    }

    private var configsPc = false
    fun getConfigsPc(): Boolean {
        return configsPc
    }

    fun setConfigsPc(configsPc: Boolean) {
        savedStateHandle.set("configsPc", configsPc)
        this.configsPc = configsPc
    }

    private var configsCc = false
    fun getConfigsCc(): Boolean {
        return configsCc
    }

    fun setConfigsCc(configsCc: Boolean) {
        savedStateHandle.set("configsCc", configsCc)
        this.configsCc = configsCc
    }

    private var configsW = false
    fun getConfigsW(): Boolean {
        return configsW
    }

    fun setConfigsW(configsW: Boolean) {
        savedStateHandle.set("configsW", configsW)
        this.configsW = configsW
    }

    private var configsL = false
    fun getConfigsL(): Boolean {
        return configsL
    }

    fun setConfigsL(configsL: Boolean) {
        savedStateHandle.set("configsL", configsL)
        this.configsL = configsL
    }

    private var configsFs = false
    fun getConfigsFs(): Boolean {
        return configsFs
    }

    fun setConfigsFs(configsFs: Boolean) {
        savedStateHandle.set("configsFs", configsFs)
        this.configsFs = configsFs
    }

    private var configsDelayWearables = 0
    fun getConfigsDelayWearables(): Int {
        return configsDelayWearables
    }

    fun setConfigsDelayWearables(configsDelayWearables: Int) {
        savedStateHandle.set("configsDelayWearables", configsDelayWearables)
        this.configsDelayWearables = configsDelayWearables
    }

    private var configsDelayLocations = 0
    fun getConfigsDelayLocations(): Int {
        return configsDelayLocations
    }

    fun setConfigsDelayLocations(configsDelayLocations: Int) {
        savedStateHandle.set("configsDelayLocations", configsDelayLocations)
        this.configsDelayLocations = configsDelayLocations
    }

    private var configsAuthfailure = 0
    fun getConfigsAuthfailure(): Int {
        return configsAuthfailure
    }

    fun setConfigsAuthfailure(configsAuthfailure: Int) {
        savedStateHandle.set("configsAuthfailure", configsAuthfailure)
        this.configsAuthfailure = configsAuthfailure
    }

    private var configsAutounlink = 0
    fun getConfigsAutounlink(): Int {
        return configsAutounlink
    }

    fun setConfigsAutounlink(configsAutounlink: Int) {
        savedStateHandle.set("configsAutounlink", configsAutounlink)
        this.configsAutounlink = configsAutounlink
    }

    private var configsAutounlinkwarning = 0
    fun getConfigsAutounlinkwarning(): Int {
        return configsAutounlinkwarning
    }

    fun setConfigsAutounlinkwarning(configsAutounlinkwarning: Int) {
        savedStateHandle.set("configsAutounlinkwarning", configsAutounlinkwarning)
        this.configsAutounlinkwarning = configsAutounlinkwarning
    }

    private var configsAllowchangesunlinked = false
    fun getConfigsAllowchangesunlinked(): Boolean {
        return configsAllowchangesunlinked
    }

    fun setConfigsAllowchangesunlinked(configsAllowchangesunlinked: Boolean) {
        savedStateHandle.set("configsAllowchangesunlinked", configsAllowchangesunlinked)
        this.configsAllowchangesunlinked = configsAllowchangesunlinked
    }

    private var authenticatorTheme: AuthenticatorTheme? = null
    fun setAuthenticatorTheme(authenticatorTheme: AuthenticatorTheme?) {
        this.authenticatorTheme = authenticatorTheme
    }

    fun finalizeChangesToConfigAndRestart(context: Context) {
        _state.value = State.ChangesLoading()
        if (configsSdkKey.trim { it <= ' ' }.isEmpty()) {
            _state.value = State.ChangesFailed("Key cannot be null or empty.")
            return
        }

        // Let Authenticator SDK validate threshold arguments
        val config: AuthenticatorConfig
        val uiConfig: AuthenticatorUIConfig
        try {
            val builder = AuthenticatorConfig.Builder()
            disallowAuthMethod(builder, AuthMethod.PIN_CODE, configsPc)
            disallowAuthMethod(builder, AuthMethod.CIRCLE_CODE, configsCc)
            disallowAuthMethod(builder, AuthMethod.WEARABLES, configsW)
            disallowAuthMethod(builder, AuthMethod.LOCATIONS, configsL)
            disallowAuthMethod(builder, AuthMethod.BIOMETRIC, configsFs)
            config = builder
                    .activationDelayWearable(configsDelayWearables)
                    .activationDelayLocation(configsDelayLocations)
                    .thresholdAuthFailure(configsAuthfailure)
                    .thresholdAutoUnlink(configsAutounlink)
                    .thresholdAutoUnlinkWarning(configsAutounlinkwarning)
                    .keyPairSize(AuthenticatorConfig.Builder.KEYSIZE_MINIMUM)
                    .sslPinning(configsSslPinning)
                    .build()
            uiConfig = AuthenticatorUIConfig.Builder()
                    .allowSecurityChangesWhenUnlinked(configsAllowchangesunlinked)
                    .theme(authenticatorTheme)
                    .build(context.applicationContext)
        } catch (e: IllegalArgumentException) {
            _state.value = State.ChangesFailed(e.message!!)
            return
        }
        if (authenticatorManager.isDeviceLinked) {
            // Force-unlink to clear any data before re-initializing
            authenticatorManager.unlinkDevice(null, object : UnlinkDeviceEventCallback() {
                override fun onSuccess(device: Device) {
                    authenticatorManager.initialize(config)
                    authenticatorUIManager.initialize(uiConfig)
                    _state.value = State.ChangesDone(configsSdkKey)
                }

                override fun onFailure(e: Exception) {
                    authenticatorManager.initialize(config)
                    authenticatorUIManager.initialize(uiConfig)
                    _state.value = State.ChangesDone(configsSdkKey)
                }
            })
        } else {
            authenticatorManager.initialize(config)
            authenticatorUIManager.initialize(uiConfig)
            _state.setValue(State.ChangesDone(configsSdkKey))
        }
    }

    private val _state: MutableLiveData<State> = SingleLiveEvent()
    val state: LiveData<State> = _state

    abstract class State {
        class ChangesFailed internal constructor(val failure: String) : State()
        class ChangesLoading : State()
        class ChangesDone internal constructor(val sdkKey: String) : State()
    }

    companion object {
        private fun disallowAuthMethod(configBuilder: AuthenticatorConfig.Builder, authMethod: AuthMethod, allow: Boolean) {
            if (!allow) {
                configBuilder.disallowAuthMethod(authMethod)
            }
        }
    }

    init {
        val currentConfig = authenticatorManager.config
        configsSdkKey = if (savedStateHandle.get<Any?>("configsSdkKey") != null) savedStateHandle.get("configsSdkKey")!! else sdkKey
        configsSslPinning = if (savedStateHandle.get<Any?>("configsSslPinning") != null) savedStateHandle.get("configsSslPinning")!! else currentConfig.isSslPinningRequired
        configsPc = if (savedStateHandle.get<Any?>("configsPc") != null) savedStateHandle.get("configsPc")!! else currentConfig.isMethodAllowed(AuthMethod.PIN_CODE)
        configsCc = if (savedStateHandle.get<Any?>("configsCc") != null) savedStateHandle.get("configsCc")!! else currentConfig.isMethodAllowed(AuthMethod.CIRCLE_CODE)
        configsW = if (savedStateHandle.get<Any?>("configsW") != null) savedStateHandle.get("configsW")!! else currentConfig.isMethodAllowed(AuthMethod.WEARABLES)
        configsL = if (savedStateHandle.get<Any?>("configsL") != null) savedStateHandle.get("configsL")!! else currentConfig.isMethodAllowed(AuthMethod.LOCATIONS)
        configsFs = if (savedStateHandle.get<Any?>("configsFs") != null) savedStateHandle.get("configsFs")!! else currentConfig.isMethodAllowed(AuthMethod.BIOMETRIC)
        configsDelayWearables = if (savedStateHandle.get<Any?>("configsDelayWearables") != null) savedStateHandle.get("configsDelayWearables")!! else currentConfig.activationDelayWearablesSeconds()
        configsDelayLocations = if (savedStateHandle.get<Any?>("configsDelayLocations") != null) savedStateHandle.get("configsDelayLocations")!! else currentConfig.activationDelayLocationsSeconds()
        configsAuthfailure = if (savedStateHandle.get<Any?>("configsAuthfailure") != null) savedStateHandle.get("configsAuthfailure")!! else currentConfig.thresholdAuthFailure()
        configsAutounlink = if (savedStateHandle.get<Any?>("configsAutounlink") != null) savedStateHandle.get("configsAutounlink")!! else currentConfig.thresholdAutoUnlink()
        configsAutounlinkwarning = if (savedStateHandle.get<Any?>("configsAutounlinkwarning") != null) savedStateHandle.get("configsAutounlinkwarning")!! else currentConfig.thresholdAutoUnlinkWarning()
        val currentUIConfig = authenticatorUIManager.config
        configsAllowchangesunlinked = if (savedStateHandle.get<Any?>("configsAllowchangesunlinked") != null) savedStateHandle.get("configsAllowchangesunlinked")!! else currentUIConfig.areSecurityChangesAllowedWhenUnlinked()
    }
}